Implement BIP 174 Partially Signed Bitcoin Transactions serialization and RPCs

achow101 opened the pull request:
[BIP 174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki) specifies a binary transaction format which contains the information necessary for a signer to produce signatures for the transaction and holds the signatures for an input while the input does not have a complete set of signatures.

***

BIP 174 is fully implemented in this pull request. It contains a struct for the a PSBT, serialization and deserialization functions, and a PSBT specific versions of `ProduceSignature` (`SignPartialTransaction`), `SignStep`(`SignSigsOnly`) and `CombineSignatures`(FinalizePartialTransaction`).

Currently PSBT functionality is limited to 4 new RPC calls, `walletupdatepsbt`, `walletcreatepsbt`, `combinepsbt`, and `decodepsbt`.

`walletupdatepsbt` updates a given PSBT with data from the wallet. For each input, it will attempt to add the proper UTXO, sign for the input, and finalize the input. It will also add any known redeem scripts, witness scripts, and public key derivation paths if they are known to the wallet.

`walletcreatepsbt` takes a network serialized raw transaction as produced by the `*rawtransaction` commands and converts it to a PSBT. Inputs are filled using information known to the wallet.

`combinepsbt` combines multiple PSBTs and finalizes them if possible.

`decodepsbt` decodes a PSBT and into a human readable format in order to more easily examine them. It is analogous to `decoderawtransaction`.

***

All of the test vectors currently in BIP 174 are also implemented.

achow101 committed:
Implement PSBT structures and un/serliazation methods per BIP 174

achow101 committed:
Implement a signer that takes an unserialized PSBT and signs it

achow101 committed:
Create RPCs for PSBT

walletupdatepsbt takes a PSBT format transaction, updates the
PSBT with any inputs related to this wallet, and signs the transaction

walletcreatepsbt takes a raw network serialized transaction like
what createrawtransaction and fundrawtransaction output and converts
it to a PSBT using whatever information about the inputs it knows
from the wallet.

decodepsbt takes a PSBT and decodes it to JSON

combinepsbt takes multiple PSBTs for the same tx and combines them.

achow101 committed:
Test the PSBT RPCs

jonasschnelli commented on src/script/sign.cpp:
```diff
@@ -302,6 +381,218 @@ struct Stacks
 }
 
+// Iterates through all inputs of the partially signed transaction and just produces signatures for what it can and adds them to the psbt partial sigs
+bool SignPartialTransaction(PartiallySignedTransaction& psbt, const CKeyStore* keystore, int nHashType)
```
`SignPartialTransaction` sounds after we have partial transactions. Suggest `SignPartialSignedTransaction`.

achow101 replied:
Changed to `SignPartiallySignedTransaction`

jonasschnelli commented:
Great work!
General concept ack, though the PR is large and maybe there is a way to make smaller steps towards BIP174 (reduce of risks).

achow101 commented:
I'm not sure what's causing the travis failure.

gmaxwell commented:
Concept ACK!

dcousens commented:
Concept ACK

laanwj commented:
Neat! Concept ACK.

instagibbs commented on src/script/sign.cpp:
```diff
@@ -302,6 +381,218 @@ struct Stacks
+                switch (whichType)
+                {
+                case TX_PUBKEY:
+                    script_ret.push_back(psbt.inputs[i].partial_sigs.find(CPubKey(vSolutions[0]))->second);
```
this may segfault when there is no corresponding `partial_sig`, I believe this is the cause of your test failure.

achow101 replied:
Fixed

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3520,6 +3520,336 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+        }
+        // Get public keys if hd is enabled
+        if (pwallet->IsHDEnabled()) {
+            if (type == TX_PUBKEYHASH) {
```
should be:

`if (type == TX_PUBKEYHASH || type == TX_WITNESS_V0_KEYHASH) {`

achow101 replied:
Done.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3520,6 +3520,336 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+            // Add to inputs
+            psbtx.inputs.push_back(input);
+        }
+    }
```
this isn't mentioned in the spec anywhere, but passing along the the change output's pubkey(s)/hdpath(s)/redeemscript allows hww to understand change outputs.

achow101 replied:
Done

achow101 replied:
I decided to make this an option instead of the default. I'll add tests for it later.

laanwj replied:
Should that be mentioned in the spec, then?

achow101 commented:
Rebased and squashed fixup commits.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4082,6 +4163,8 @@ UniValue walletcreatepsbt(const JSONRPCRequest& request)
                             "using fundrawtransaction.\n"
                             "\nArguments:\n"
                             "1. \"hexstring\"            (string, required) The hex string of the raw transaction\n"
+                            "2. \"include_output_info\"  (boolean, optional, default=false) If true, returns the PSBT with the redeem scripts, witness\n"
```
you need to edit number of allowed args, and include the optional one in parens above.

instagibbs replied:
you also need to add it to the list of arguments that will be parsed as json rather than strings in `rpc/client.cpp`

achow101 replied:
Dang, I knew I forgot something. Fixed.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3956,6 +3956,84 @@ void fill_psbt(const CWallet* pwallet, PartiallySignedTransaction& psbtx, const
+                if (type == TX_PUBKEYHASH || type == TX_WITNESS_V0_KEYHASH) {
+                    uint160 hash(solns[0]);
+                    CKeyID keyID(hash);
+                    if (!pwallet->HaveKey(keyID)) {
```
What we really care about is that we have the hdkeypath, not particularly that the key is present or not. I don't think these `HaveKey` checks are necessary since `add_keypath_to_map` will deal with that case specifically.

achow101 replied:
Right, fixed.

instagibbs commented on src/rpc/client.cpp:
```diff
@@ -102,6 +102,8 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "fundrawtransaction", 2, "iswitness" },
     { "walletupdatepsbt", 2, "psbtformat"},
+    { "walletupdatepsbt", 3, "include_output_info"},
+    { "walletcreatepsbt", 2, "include_output_info"},
```
should be `1`

achow101 replied:
fixed

instagibbs reviewed (commented):
"ERROR: walletcreatepsbt argument 2 (named include_output_info in vRPCConvertParams) is not defined in dispatch table
ERROR: walletupdatepsbt argument 3 (named include_output_info in vRPCConvertParams) is not defined in dispatch table"

You need to add the new named args to the dispatch table

instagibbs commented:
like tACK

I have rebased the externalhd branch onto this PR, with minor modifications, for ledger support, combined with @achow101 's HWI repo for signing. https://github.com/instagibbs/bitcoin/tree/external_psbt

instagibbs commented:
needs rebase

achow101 commented:
Rebased

achow101 commented:
Rebased

sipa commented on src/script/sign.h:
```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+                        s >> tx;
+                    } else {
+                        // Read in the transaction
+                        CMutableTransaction mtx;
```
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

You can use `CTransactionRef prev_tx; s >> prev_tx;` instead here (and `input.non_witness_utxo = std::move(prev_tx);` below).

sipa commented on src/script/sign.h:
```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
```
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

General comment on deserialization code: check that keys and values (where relevant) have the expected length. Things like `uint160(key.begin() + 1, key.end())` will assertion fail if the length is wrong.

sipa reviewed (commented):
First set of comments on serialization code.

achow101 commented on src/script/sign.h:
```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+
+        // Write the number of inputs
+        if (num_ins > 0) {
+            s << PSBT_NUM_IN_VIN;
```
It specifies the number of inputs that have input data in the PSBT, not the number of inputs in the transaction in general.

sipa replied:
@achow101 Ok, makes sense.

But you're still writing PSB_NUM_IN_VIN without a keylength prefix here.

achow101 commented on src/script/sign.h:
```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+                if (use_in_index) {
+                    WriteCompactSize(s, sizeof(PSBT_NUM_IN_VIN));
+                    s << PSBT_NUM_IN_VIN;
+                    WriteCompactSize(s, sizeof(psbt_in.index));
```
> This is a lot easier if you'd have a single-argument SerializeToVector first (see above).

Above where?

I'm not necessarily sure that that will always work due to the different data types.

> Perhaps it's generally better to use the separator approach instead?

I think it generally is better to use the separator approach. By default, this code does not use input indexes unless a psbt that uses input indexes was provided. That option was provided at the suggestion of someone else in order to handle very large transactions where almost all inputs were signed.

sipa replied:
> Above where?

https://github.com/bitcoin/bitcoin/pull/12136/commits/e25f290841de2c446a237ad3942a0c2c0427ee62#r184790215

> I'm not necessarily sure that that will always work due to the different data types.

It should. Char vectors are always serialized as compactsize-in-bytes + data-bytes. SerializeToVector first serializes to a vector, and then you serialize that. I can create a small example if you want.

This is still writing the input index as a 4-byte int, rather than as a compactsize, btw.

achow101 commented:
Addressed all of @sipa's comments.

achow101 commented on src/script/sign.h:
```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+        // Write any hd keypaths
+        for (auto keypath_pair : hd_keypaths) {
+            s << SerializeToVector(PSBT_BIP32_KEYPATH_SIGHASH, Span<const unsigned char>(keypath_pair.first.begin(), keypath_pair.first.size()));
+            WriteCompactSize(s, keypath_pair.second.size() * sizeof(uint32_t));
```
This does not work because it will be prefixed with the number of items in the keypath, not with the total size in bytes that that will be. Each item in the keypath is a uint32_t, not an unsigned char.

sipa commented on src/script/sign.h:
```diff
@@ -68,6 +71,405 @@ struct SignatureData {
+                // Write any partial signatures
+                for (auto sig_pair : psbt_in.partial_sigs) {
+                    s << SerializeToVector(PSBT_WITNESSSCRIPT_PARTIAL_SIG, Span<const unsigned char>(sig_pair.first.begin(), sig_pair.first.size()));
+                    WriteCompactSize(s, sig_pair.second.size());
```
This and the line below can be written as 's << sig_pair.second`, I think - it's just an unsigned char vector.

sipa commented:
@achow101 Here's a commit with a few suggested improvements to the serialization code: https://github.com/sipa/bitcoin/commit/8733102e713b1ef8e96f6fca80e77a5d8c47fc0b

achow101 commented:
I incorporated @sipa's serialization improvements.

instagibbs commented:
needs rebase

achow101 commented:
Rebased

achow101 commented:
Rebased this and did some reorganization.

I squashed down a lot of the commits so that the changes to different parts aren't scattered around in different commits.

I also de-duplicated the signing code by using a PSBT specific SIgningProvider and SignatureCreator. The SigningProvider handles providing pubkeys and scripts from the PSBT itself along with the wallet and the SignatureCreator adds signatures and public keys directly to the PSBT once the signature is created. This avoids the previous code duplication.

Hopefully the reorganization and the PSBT specific SigningProvider and SignatureCreator will make this easier to review.

laanwj commented on src/uint256.h:
```diff
@@ -91,6 +91,15 @@ class base_blob
 
+    uint32_t GetUint32(int pos) const
+    {
+        const uint8_t* ptr = data + pos * 8;
```
why *8? that doesn't look in line with 32 bit values.

Also, please add `LE` to make clear that this is always little-endian.

kallewoof replied:
Maybe I am seeing outdated code, but it still says `* 8` here and `uint64_t` below.

laanwj commented on test/functional/test_runner.py:
```diff
@@ -145,6 +145,7 @@
     'feature_blocksdir.py',
     'feature_config_args.py',
     'feature_help.py',
+    "rpc_psbt.py",
```
Single quote like the others?

dcousens commented:
@achow101 have you thought about making this a 3rd party library for integration into other software,  and then integration here?

jeffrade commented on test/functional/test_runner.py:
```diff
@@ -145,6 +145,7 @@
     'feature_blocksdir.py',
     'feature_config_args.py',
     'feature_help.py',
+    'rpc_psbt.py',
```
```
rpc_psbt.py                           | ✓ Passed  | 4 s
.
.
.
rpc_users.py                          | ✓ Passed  | 3 s
```

Since `rpc_psbt.py` ran locally for me in 4 seconds and to avoid possible merge conflict, `'rpc_psbt.py'` can be inserted before `'rpc_users.py'` in this list.

achow101 replied:
Done

kallewoof commented on src/script/sign.h:
```diff
@@ -68,6 +71,432 @@ struct SignatureData {
+                        unsigned char rs_hash160_data[hash160_hasher.OUTPUT_SIZE];
+                        hash160_hasher.Finalize(rs_hash160_data);
+                        uint160 rs_hash160(std::vector<unsigned char>(rs_hash160_data, rs_hash160_data + hash160_hasher.OUTPUT_SIZE));
+                        if (hash160.Compare(rs_hash160) != 0) {
```
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

Doesn't really make a huge difference, but I think `if (hash160 != rs_hash160)` is a tiny bit more readable.

achow101 replied:
Done

kallewoof commented on src/script/sign.h:
```diff
@@ -68,6 +71,432 @@ struct SignatureData {
+                        unsigned char ws_sha256_data[sha256_hasher.OUTPUT_SIZE];
+                        sha256_hasher.Finalize(ws_sha256_data);
+                        uint256 ws_sha256(std::vector<unsigned char>(ws_sha256_data, ws_sha256_data + sha256_hasher.OUTPUT_SIZE));
+                        if (hash.Compare(ws_sha256) != 0) {
```
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

Same here about `(hash != ws_sha256)`.

achow101 replied:
Done

kallewoof commented on src/script/sign.h:
```diff
@@ -68,6 +71,432 @@ struct SignatureData {
+                    break;
+                // BIP 32 HD Keypaths and sighash types
+                case PSBT_BIP32_KEYPATH_SIGHASH:
+                {
```
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

Why the `{}` here and in `case PSBT_NUM_IN_VIN:` below? No vars are declared.

achow101 replied:
Removed

kallewoof commented on src/script/sign.h:
```diff
@@ -68,6 +71,432 @@ struct SignatureData {
+                            throw std::ios_base::failure("Size of key was not the expected size for the type BIP32 keypath");
+                        }
+                        // Read in the pubkey from key
+                        CPubKey pubkey(key.begin() + 1, key.end());
```
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

Might be good to check validity of the pubkey. E.g.
```C++
if (!pubkey.IsValid()) {
  throw std::ios_base::failure("Invalid pubkey");
}
```

instagibbs replied:
Might want `IsFullyValid()`

achow101 replied:
Done

kallewoof commented on src/script/sign.cpp:
```diff
@@ -33,6 +33,23 @@ bool TransactionSignatureCreator::CreateSig(const SigningProvider& provider, std
+bool PSBTSignatureCreator::CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& address, const CScript& scriptCode, SigVersion sigversion) const
+{
+    CPubKey pubkey;
+    provider.GetPubKey(address, pubkey);
```
In commit "Implement a signer that takes an unserialized PSBT and signs it":

Check result of `provider.GetPubKey()` and react if `false`?

achow101 replied:
Done

kallewoof commented on src/script/sign.cpp:
```diff
@@ -305,6 +322,118 @@ struct Stacks
+    // Look for scripts in redeem_scripts
+    auto mi = psbt->redeem_scripts.find(scriptid);
+    if (mi != psbt->redeem_scripts.end())
+    {
```
Nit/style: `{` on same line as `if`

achow101 replied:
Done

kallewoof commented on src/script/sign.h:
```diff
@@ -497,9 +498,35 @@ struct PartiallySignedTransaction
 /** Produce a script signature using a generic signature creator. */
 bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreator& creator, const CScript& scriptPubKey, SignatureData& sigdata);
 
+bool SignPartiallySignedTransaction(PartiallySignedTransaction& psbt, SigningProvider* provider, int nHashType, bool finalize = false);
```
In commit "Implement a signer that takes an unserialized PSBT and signs it":

Maybe add a comment about this function, in particular saying that it returns true if it was complete, and if `finalize` was set, and `false` does not necessarily mean failure to partially sign.

achow101 replied:
Done

kallewoof commented on src/uint256.h:
```diff
@@ -91,13 +91,13 @@ class base_blob
     }
 
-    uint32_t GetUint32(int pos) const
+    uint32_t GetUint32LE(int pos) const
```
In commit "Implement a signer that takes an unserialized PSBT and signs it":

Was this meant to go into the previous commit?

achow101 replied:
Oops, yes. Fixed

kallewoof reviewed (commented):
Partially reviewed:

- [x] c3d7c855d239943fb85a2fce1c5dae1a3f999a71 Implement PSBT structures and un/serliazation methods per BIP 174
- [x] ed1a6e3ea2ed526dab76f15ec71b2f7031bd2335 Implement a signer that takes an unserialized PSBT and signs it
- [ ] ac2b5093282e66a4ec1a579fb0ea133e9880ae08 Create RPCs for PSBT
- [ ] 92cca779d6604762c0e2ee4bcd61dc17502e12f4 Test the PSBT RPCs

instagibbs commented on src/rpc/client.cpp:
```diff
@@ -110,6 +110,11 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "combinerawtransaction", 0, "txs" },
     { "fundrawtransaction", 1, "options" },
     { "fundrawtransaction", 2, "iswitness" },
+    { "walletupdatepsbt", 2, "psbtformat"},
```
I think the `wallet` prefixes are helpful

pedrobranco replied:
> I think the wallet prefixes are helpful

>I prefer the wallet prefixes and expanding out psbt to partiallysignedtransaction just makes the RPC names too long IMO.

We now have several wallet RPC methods that are not prefixed with `wallet`, so IMO is an unnecessary prefix. 

Also, by the spec `PSBT` is a transaction format, so the transaction itself is a `partially signed transaction`, or `pst`, so using `psbt` as suffix somehow seems improper.

Given that, WDYT about not using the prefix `wallet`, use the `ps` (term for `partially signed`) as a transaction type, and use the suffix `transaction`:
- `walletcreatepsbt` to `createpstransaction`
- `walletupdatepsbt` to `updatepstransaction`
- `combinepsbt` to `combinepstransactions`
- `decodepsbt` to `decodepstransaction`

sipa replied:
I think it's the other way around. Historically no RPCs had a wallet prefix, which led to a lot of confusion (e.g. gettransaction vs getrawtransaction).

I'm in favor of having wallet in the name of RPCs that require a wallet to interact with.

dcousens replied:
`bip174-create`
`bip174-update`
`bip174-combine`
`bip174-decode`

achow101 replied:
I agree with @sipa and prefer to have the wallet prefix, especially because there will probably be a non-wallet version of those RPCs at some point in the future.

I don't think psbt is really that ambiguous..

pedrobranco replied:
> I think it's the other way around. Historically no RPCs had a wallet prefix, which led to a lot of confusion (e.g. gettransaction vs getrawtransaction).

Agree on that. If all methods will be prefixed LGTM.

achow101 commented on src/script/sign.h:
```diff
@@ -68,6 +71,432 @@ struct SignatureData {
+    CVectorWriter ss(SER_NETWORK, PROTOCOL_VERSION, ret, 0);
+    SerializeMany(ss, args...);
+    return ret;
+}
```
Not sure, it's kind of specific to PSBT.

kallewoof replied:
Looks very generic, but I have no strong feelings about it (commit still has typo btw).

achow101 commented on src/script/sign.h:
```diff
@@ -68,6 +71,432 @@ struct SignatureData {
+        }
+
+        // Write the unknown things
+        for(auto& entry : unknown) {
```
Done

achow101 commented on src/core_write.cpp:
```diff
@@ -70,6 +70,15 @@ const std::map<unsigned char, std::string> mapSigHashTypes = {
+{
+    if (mapSigHashTypes.count(sighash_type)) {
+        return mapSigHashTypes.find(sighash_type)->second;
+    } else {
```
Removed

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4292,6 +4292,408 @@ UniValue sethdseed(const JSONRPCRequest& request)
+            if (first) {
+                first = false;
+                continue;
+            } else {
```
Removed

achow101 commented:
@dcousens I have, but I'm not sure what the third party library would really be.

kallewoof commented on src/script/sign.h:
```diff
@@ -68,6 +71,434 @@ struct SignatureData {
+                case PSBT_BIP32_KEYPATH_SIGHASH:
+                    if (in_globals) {
+                        // Make sure that the key is the size of pubkey + 1
+                         if (key.size() != CPubKey::PUBLIC_KEY_SIZE + 1 && key.size() != CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1) {
```
In "Implement PSBT structures and un/serliazation methods per BIP 174":

Nit: mis-aligned indentation. (remove 1 space I think?)

achow101 replied:
Done

kallewoof commented on src/script/sign.h:
```diff
@@ -68,6 +71,434 @@ struct SignatureData {
+                    std::vector<unsigned char> val_bytes;
+                    val_bytes.resize(value_len);
+                    Span<unsigned char> value = MakeSpan(val_bytes);
+                    s >> value;
```
In "Implement PSBT structures and un/serliazation methods per BIP 174":

I think this (and other places where `Span` is used) would look less confusing if you did `s >> MakeSpan(val_bytes)`, since you are not using `value`, and it looks like the data is written to some unrelated thing (i.e. not `val_bytes`, but `value`). Not sure if that is possible though (it should be, if it isn't, but that's a separate story).

achow101 replied:
Done here and everywhere else this happens.

kallewoof reviewed (approved):
utACK 30c0cf203e8c82798e047ab4b8805638612421a5

laanwj reviewed (commented):
some comments, I did not get around to reviewing all changes yet

achow101 commented on src/script/sign.h:
```diff
@@ -68,6 +71,434 @@ struct SignatureData {
+    return ret;
+}
+
+/** An structure for PSBTs which contain per input information */
```
Done

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,422 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+UniValue decodepsbt(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 1)
+    throw std::runtime_error(
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4292,6 +4292,408 @@ UniValue sethdseed(const JSONRPCRequest& request)
+        throw std::runtime_error(
+            "walletupdatepsbt \"hexstring\" ( sighashtype psbtformat include_output_info)\n"
+            "\nUpdate a psbt with input information from our wallet and then sign inputs\n"
+            "that we an sign for.\n"
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4292,6 +4292,408 @@ UniValue sethdseed(const JSONRPCRequest& request)
+            "3. \"psbtformat\"              (boolean, optional, default=false) If true, return the complete transaction \n"
+            "                             in the PSBT format. Otherwise complete transactions will be in normal network serialization.\n"
+            "4. \"include_output_info\"     (boolean, optional, default=false) If true, returns the PSBT with the redeem scripts, witness\n"
+            "                             scripts, and bip32 keypaths of the outputs if they are available. This is useful for hardware wllets\n"
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4292,6 +4292,408 @@ UniValue sethdseed(const JSONRPCRequest& request)
+
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 4)
+        throw std::runtime_error(
+            "walletupdatepsbt \"hexstring\" ( sighashtype psbtformat include_output_info)\n"
```
Done

achow101 commented on src/core_write.cpp:
```diff
@@ -70,6 +70,14 @@ const std::map<unsigned char, std::string> mapSigHashTypes = {
 
+std::string SighashToStr(unsigned char sighash_type)
+{
+    if (mapSigHashTypes.count(sighash_type)) {
```
Done

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,422 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
     return result;
 }
 
+void write_hd_keypath(std::vector<uint32_t>& keypath, std::string& keypath_str)
```
Done

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,422 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+        "\nResult:\n"
+        "{\n"
+        "  \"tx\" : {                   (json object) The decoded network serialized transaction\n"
+        "    ...                                      decoding is the same as decoderawtransaction's."
```
Done

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,422 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+        "      \"hash\" : \"hash\",       (string) The hash160 of the redeem script\n"
+        "      \"script\" : {           (json object)\n"
+        "        \"asm\":\"asm\",         (string) Script public key\n"
+        "        \"hex\":\"hex\",         (string) hex encoded public key\n"
```
Done

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,422 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            keypath.pushKV("pubkey", HexStr(entry.first));
+
+            uint32_t fingerprint = entry.second.at(0);
+            std::vector<uint8_t> fingerprint_bytes;
```
Done

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,422 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+        if (!ssData.empty()) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed, extra data after PSBT");
+        }
+    }
```
Done

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,422 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            keypath.pushKV("master_fingerprint", HexStr(fingerprint_bytes));
+
+            std::string keypath_str;
+            entry.second.erase(entry.second.begin());
```
Since entry is not a reference or a pointer, `psbtx` isn't actually modified.

achow101 commented:
Also rebased to (hopefully) fix the travis error

promag commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,415 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            + HelpExampleRpc("decodepsbt", "\"hexstring\"")
+    );
+
+    RPCTypeCheck(request.params, {UniValue::VSTR});
```
I believe this can be removed, `get_str()` below will validate it.

promag commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1259,415 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    if (!IsHex(request.params[0].get_str())) {
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed, not hex");
+    }
+    std::vector<unsigned char> txData(ParseHex(request.params[0].get_str()));
```
Use `ParseHexV` instead of `IsHex` + `ParseHex`?

achow101 replied:
Done

promag reviewed (commented):
This is a lot to review and will probably receive a lot of more comments. Have you thought of splitting it?

promag commented on src/script/sign.h:
```diff
@@ -61,6 +64,429 @@ struct SignatureData {
+    friend bool operator==(const PartiallySignedTransaction& a, const PartiallySignedTransaction &b)
+    {
+        // Check the inputs
+        for (unsigned int i = 0; i < a.tx.vin.size(); ++i) {
```
First check?
```cpp
if (a.tx.vin.size() != b.tx.vin.size() || a.tx.vout.size() != b.tx.vout.size()) return false;
```
Otherwise this allows `b` to have more in/outputs and if `b` has fewer then it oob.

achow101 replied:
Done

achow101 commented on src/core_write.cpp:
```diff
@@ -70,6 +70,25 @@ const std::map<unsigned char, std::string> mapSigHashTypes = {
 
+std::string SighashToStr(unsigned char sighash_type)
+{
+    auto it = mapSigHashTypes.find(sighash_type);
```
Done

achow101 commented on src/core_write.cpp:
```diff
@@ -70,6 +70,25 @@ const std::map<unsigned char, std::string> mapSigHashTypes = {
+std::string SighashToStr(unsigned char sighash_type)
+{
+    auto it = mapSigHashTypes.find(sighash_type);
+    if (it != mapSigHashTypes.end()) {
```
Done

promag replied:
Missing change here and above.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4292,6 +4292,408 @@ UniValue sethdseed(const JSONRPCRequest& request)
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 4)
+        throw std::runtime_error(
+            "walletupdatepsbt \"hexstring\" ( sighashtype psbtformat include_output_info )\n"
+            "\nUpdate a psbt with input information from our wallet and then sign inputs\n"
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4292,6 +4292,408 @@ UniValue sethdseed(const JSONRPCRequest& request)
+        if (!ssData.empty()) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed, extra data after PSBT");
+        }
+    }
```
Done

achow101 commented:
> Have you thought of splitting it?

Splitting into what components? All of it kind of goes together.

pedrobranco commented on src/script/sign.cpp:
```diff
@@ -304,6 +323,119 @@ struct Stacks
+        solved &= sig_complete;
+    }
+
+    return solved;
```
Keep the code format consistent (other places without new lines before `return`). 

IMHO we should have some set of rules for code formatting (maybe a linter would be nice).

achow101 replied:
IMO this doesn't matter

achow101 commented:
Rebased to fix a silent merge conflict.

promag commented on src/core_write.cpp:
```diff
@@ -70,6 +70,25 @@ const std::map<unsigned char, std::string> mapSigHashTypes = {
+
+std::vector<uint8_t> Uint32ToUint8VectorLE(uint32_t in)
+{
+    std::vector<uint8_t> bytes;
```
Alternative
```cpp
return { uint8_t(in), uint8_t(in >> 8), uint8_t(in >> 16), uint8_t(in >> 24) };
```

tokutech commented:
Hello,

Why not using JSON(like protocolbuffer) instead this BIP format? Its more simpler.

achow101 commented:
@tokutech Because stuff like JSON or protobufs are more cumbersome to use. For JSON, the size of  the objects  are much larger. They are also in a form that is human readable and human readability doesn't matter for this.  Furthermore, in order for devices to use something that is JSON formatted, it needs to have a JSON parser, which is not necessarily something that exists on such devices. While those devices need to have a PSBT parser to use PSBT, that is easier as much of the functionality required for a PSBT parser is already required to parse other Bitcoin network messages, including normal transactions.

As for protobufs, that introduces another dependency and we don't want to do that.

achow101 committed:
Make SignatureData able to store signatures and scripts

In addition to having the scriptSig and scriptWitness, have SignatureData
also be able to store just the signatures (pubkeys mapped to sigs) and
scripts (script ids mapped to scripts).

Also have DataFromTransaction be able to extract signatures and scripts
from the scriptSig and scriptWitness of an input to put them in SignatureData.

Adds a new SignatureChecker which takes a SignatureData and puts pubkeys
and signatures into it when it successfully verifies a signature.

Adds a new field in SignatureData which stores whether the SignatureData
was complete. This allows us to also update the scriptSig and
scriptWitness to the final one when updating a SignatureData with another
one.

achow101 committed:
Replace CombineSignatures with ProduceSignature

Instead of using CombineSignatures to create the final scriptSig or
scriptWitness of an input, use ProduceSignature itself.

To allow for ProduceSignature to place signatures, pubkeys, and scripts
that it does not know about, we pass down the SignatureData to SignStep
which pulls out the information that it needs from the SignatureData.

achow101 committed:
Remove CombineSignatures and replace tests

Removes CombineSignatures and replaces its use in tests with
ProduceSignature to test the same behavior for ProduceSignature.

achow101 committed:
Inline Sign1 and SignN

Sign1 and SignN are kind of redundant so remove them and inline their
behavior into SignStep

achow101 commented:
Rebased on top of #13425 which simplifies the signer logic.

sipa commented on src/script/sign.cpp:
```diff
@@ -86,16 +92,21 @@ static bool SignStep(const SigningProvider& provider, const BaseSignatureCreator
+        CPubKey pubkey(vSolutions[0]);
+        keyID = pubkey.GetID();
+        if (!creator.CreateSig(provider, sig, keyID, scriptPubKey, sigversion)) return false;
+        sigdata.signatures.emplace(keyID, SigPair(pubkey, sig));
```
Why is this needed? Signing a TX_PUBKEY should always result in a full signature.

achow101 replied:
A PSBT does not have to create the final scriptSig even when a complete set of signatures is available. The user has the option to not finalize the scriptSigs even when they can be.

sipa replied:
In that case you should also be able to deal with unfinalized signatures already present, and turn them into scriptSigs. In other words, you need to test whether sigdata.signatures contains a matching signature and reuse it. If that's the case, we're probably better off reintroducing SignatureDataSignatureCreator and SignatureDataSigningProvider, so these keys/script/signatures are automatically available everywhere.

Another interpretation is that TX_PUBKEY and TX_PUBKEYHASH etc just don't have "partial" forms - they're either fully signed or not, and as a result there never is anything to finalize.

achow101 replied:
We should discuss this on the mailing list as it requires changing the BIP.

For now, I have added a helper function for createSig which fetches and places signatures to and from the SignatureData.

sipa commented on src/script/sign.cpp:
```diff
@@ -86,16 +92,21 @@ static bool SignStep(const SigningProvider& provider, const BaseSignatureCreator
-        ret.push_back(std::move(sig));
         CPubKey pubkey;
         GetPubKey(&provider, &sigdata, keyID, pubkey);
+        sigdata.signatures.emplace(keyID, SigPair(pubkey, sig));
```
Same here.

araspitzu commented on src/script/sign.h:
```diff
@@ -53,26 +59,456 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
+                    if (in_globals) {
+                        num_ins = ReadCompactSize(s);
+                    } else {
+                        // Make sure that we are using inout indexes or this is the first input
```
small typo: Make sure that we are using input indexes or this is the first input

achow101 replied:
Fixed

achow101 commented:
Updated the title to indicate that this is primarily serializations and RPCs.

luke-jr commented:
If the only reason not to use protobufs is "another dependency", that's a bad reason to avoid it. Dependencies are better than reinventing the same thing... no reason to avoid them.

luke-jr commented:
(not to mention that protobuf is *already* a dependency)

luke-jr commented:
@sipa My goal was not to argue for a change in BIP 174, only to get a better explanation/rationale for why protobufs aren't used. :)

achow101 commented:
Since BIP 174 is changing soon, I'm closing this for now.
